node/core/provisioner/src/lib.rs

```diff
@@ -700,42 +658,98 @@ fn extend_by_random_subset_without_repetition(
 	acc.sort_unstable_by(|a, b| a.0.cmp(&b.0));
 }
 
+/// The maximum number of disputes Provisioner will include in the inherent data.
+/// Serves as a protection not to flood the Runtime with excessive data.
+const MAX_DISPUTES_FORWARDED_TO_RUNTIME: usize = 1_000;
+
 async fn select_disputes(
 	sender: &mut impl SubsystemSender,
 	metrics: &metrics::Metrics,
+	_leaf: &ActivatedLeaf,
 ) -> Result<MultiDisputeStatementSet, Error> {
-	const MAX_DISPUTES_FORWARDED_TO_RUNTIME: usize = 1_000;
-
 	// We use `RecentDisputes` instead of `ActiveDisputes` because redundancy is fine.
 	// It's heavier than `ActiveDisputes` but ensures that everything from the dispute
 	// window gets on-chain, unlike `ActiveDisputes`.
 	// In case of an overload condition, we limit ourselves to active disputes, and fill up to the
 	// upper bound of disputes to pass to wasm `fn create_inherent_data`.
 	// If the active ones are already exceeding the bounds, randomly select a subset.
 	let recent = request_disputes(sender, RequestType::Recent).await;
+
+	// On chain disputes are fetched from the runtime. We want to prioritise the inclusion of unknown
+	// disputes in the inherent data. The call relies on staging Runtime API. If the staging API is not
+	// enabled in the binary an empty set is generated which doesn't affect the rest of the logic.
+	let onchain = match onchain_disputes::get_onchain_disputes(sender, _leaf.hash.clone()).await {
+		Ok(r) => r,
+		Err(e) => {
+			gum::warn!(
```

Comment:
this will spam mainnets
please downgrade to debug

Reply:
True.
Staging API makes error handling very confusing. While this behind the feature flag - this warn will never be shown, especially on main net. But at the moment the function is promoted - it will spam the mainnet as you said.

Reply:
This is why we need to do at-runtime branching based on runtime API version.